src/multisig/multisig.cpp

```diff
@@ -57,134 +75,192 @@ namespace cryptonote
+    // 2. insert to final keyset if not there yet
+    // 3. save the corresponding index of input priv key set for later reference
+    std::unordered_map<crypto::public_key, std::size_t> own_keys_mapping;
 
-    for (const auto &k: derivations)
+    for (std::size_t multisig_keys_index{0}; multisig_keys_index < privkeys_inout.size(); ++multisig_keys_index)
     {
-      multisig_keys.emplace_back(get_multisig_blinded_secret_key(rct::rct2sk(rct::pk2rct(k))));
+      crypto::public_key pubkey;
+      CHECK_AND_ASSERT_THROW_MES(crypto::secret_key_to_public_key(privkeys_inout[multisig_keys_index], pubkey), "Failed to derive public key");
+
+      own_keys_mapping[pubkey] = multisig_keys_index;
+
+      // input should be sanitized (no duplicates)
+      CHECK_AND_ASSERT_THROW_MES(std::find(mutable_keys.begin(), mutable_keys.end(), pubkey) == mutable_keys.end(),
+        "Unexpected duplicate found in input list(s).");
+
+      mutable_keys.push_back(pubkey);
     }
 
-    return multisig_keys;
-  }
-  //-----------------------------------------------------------------
-  crypto::secret_key generate_multisig_view_secret_key(const crypto::secret_key &skey, const std::vector<crypto::secret_key> &skeys)
-  {
-    crypto::secret_key view_skey = get_multisig_blinded_secret_key(skey);
-    for (const auto &k: skeys)
-      sc_add((unsigned char*)&view_skey, rct::sk2rct(view_skey).bytes, rct::sk2rct(k).bytes);
-    return view_skey;
-  }
-  //-----------------------------------------------------------------
-  crypto::public_key generate_multisig_M_N_spend_public_key(const std::vector<crypto::public_key> &pkeys)
-  {
-    rct::key spend_public_key = rct::identity();
-    for (const auto &pk: pkeys)
+    // sort input final keys for computing aggregation coefficients (lowest to highest)
+    sort_multisig_shares_for_aggregation_coefficient(mutable_keys);
+
+    // key aggregation
+    rct::key aggregate_key = rct::identity();
+
+    for (const auto &key : mutable_keys)
     {
-      rct::addKeys(spend_public_key, spend_public_key, rct::pk2rct(pk));
+      // get aggregation coefficient
+      rct::key coeff = compute_multisig_aggregation_coefficient(mutable_keys, key);
+
+      // convert private key if possible
+      // note: retain original priv key index in input list, in case order matters upstream
+      auto found_key = own_keys_mapping.find(key);
+      if (found_key != own_keys_mapping.end())
+      {
+        // k_agg = coeff*k_base
+        sc_mul((unsigned char*)&(privkeys_inout[found_key->second]),
+          coeff.bytes,
+          (const unsigned char*)&(privkeys_inout[found_key->second]));
+
+        CHECK_AND_ASSERT_THROW_MES(privkeys_inout[found_key->second] != crypto::null_skey,
+          "Multisig privkey with aggregation coefficient unexpectedly null.");
+      }
+
+      // convert public key (pre-merge operation)
+      // K_agg = coeff*K_base
+      rct::key converted_pubkey = rct::scalarmultKey(rct::pk2rct(key), coeff);
```

Comment:
Since the `rct` library is a lot larger than `crypto`, I decided to use `rct` as much as possible. I only use `crypto` for objects that need to be memsafe (`crypto::secret_key`, `crypto::public_key_memsafe`) or if it's required by another library call.

Improving `crypto` is out of scope for this PR, and probably impractical at this point given the ubiquity of `rct`.